compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -170,28 +170,30 @@ impl DeepRejectCtxt {
     pub fn args_may_unify<'tcx>(
         self,
         obligation_args: GenericArgsRef<'tcx>,
-        impl_args: GenericArgsRef<'tcx>,
+        candidate_args: GenericArgsRef<'tcx>,
     ) -> bool {
-        iter::zip(obligation_args, impl_args).all(|(obl, imp)| {
+        iter::zip(obligation_args, candidate_args).all(|(obl, imp)| {
             match (obl.unpack(), imp.unpack()) {
                 // We don't fast reject based on regions.
                 (GenericArgKind::Lifetime(_), GenericArgKind::Lifetime(_)) => true,
-                (GenericArgKind::Type(obl), GenericArgKind::Type(imp)) => {
-                    self.types_may_unify(obl, imp)
+                (GenericArgKind::Type(obl), GenericArgKind::Type(candidate)) => {
+                    self.types_may_unify(obl, candidate)
                 }
-                (GenericArgKind::Const(obl), GenericArgKind::Const(imp)) => {
-                    self.consts_may_unify(obl, imp)
+                (GenericArgKind::Const(obl), GenericArgKind::Const(candidate)) => {
+                    self.consts_may_unify(obl, candidate)
                 }
                 _ => bug!("kind mismatch: {obl} {imp}"),
             }
         })
     }
 
-    pub fn types_may_unify<'tcx>(self, obligation_ty: Ty<'tcx>, impl_ty: Ty<'tcx>) -> bool {
-        match impl_ty.kind() {
-            // Start by checking whether the type in the impl may unify with
+    pub fn types_may_unify<'tcx>(self, obligation_ty: Ty<'tcx>, candidate_ty: Ty<'tcx>) -> bool {
+        match candidate_ty.kind() {
+            // Start by checking whether the type in the candidate may unify with
             // pretty much everything. Just return `true` in that case.
-            ty::Param(_) | ty::Error(_) | ty::Alias(..) => return true,
+            ty::Param(_) | ty::Error(_) | ty::Alias(..) | ty::Infer(_) | ty::Placeholder(..) => {
```

Comment:
This treats params always as bound vars (which is true for unsubstituted impl headers), vs when they come from the environment they should be treated as placeholders (param-env candidate matching).

I think drcx should be reworked to make this distinction, so we don't return `true` more than we need to.

Reply:
Ideally this could then be used when matching param-env candidates in the new and old solvers, so we can avoid instantiating binders and doing type equality when it's useless.